packages/@cdktf/hcl2cdk/lib/expressions.ts

```diff
@@ -175,25 +175,46 @@ export function extractReferencesFromExpression(
   }, [] as Reference[]);
 }
 
-export function referenceToVariableName(ref: Reference): string {
+export function referenceToVariableName(scope: Scope, ref: Reference): string {
   const parts = ref.referencee.full.split(".");
   const resource = parts[0] === "data" ? `${parts[0]}.${parts[1]}` : parts[0];
   const name = parts[0] === "data" ? parts[2] : parts[1];
-  return variableName(resource, name);
+  return variableName(scope, resource, name);
 }
 
-export function variableName(resource: string, name: string): string {
-  const proposedName = camelCase(
-    ["var", "local", "module"].includes(resource)
-      ? name
-      : [resource, name].join("_")
-  );
+function validVarName(name: string) {
+  if (!Number.isNaN(parseInt(name[0], 10))) {
+    return `d${name}`;
+  } else {
+    return name;
+  }
+}
+
+export function variableName(
+  scope: Scope,
+  resource: string,
+  name: string
+): string {
+  // name collision, we need to prefix the name
+  if (scope.variables[name]) {
+    if (resource === scope.variables[name].resource) {
+      return scope.variables[name].variableName;
+    }
 
-  if (!Number.isNaN(parseInt(proposedName[0], 10))) {
-    return `d${proposedName}`;
+    // we only cache one per name
+    return validVarName(camelCase([resource, name].join("_")));
```

Comment:
Technically this could still collide, but probably don't need to worry about that.